File: 01_Array_And_Hashing/two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // create empty hashmap
    const hashmap = {};

    // loop list via its length
    for (let index = 0; index < nums.length; index++) {
        const num = nums[index];          // get each number from list
        const diff = target - num;    // target minus the value in each list
        
        // check if diff is in hashmap
        if (hashmap.hasOwnProperty(diff)) {
            return [hashmap[diff], index];
        }
        
        // update hashmap -> list value: list index
        hashmap[num] = index;
    }
    return null
};


// implementation
console.log("Leetcode 1: Two Sum\n")
// Example 1
const example1 = [2, 7, 11, 15];
const target1 = 9;
console.log(`Example 1: nums = [${example1}], target = ${target1}`)
console.log("Output:", twoSum(example1, target1))

// Example 2
const example2 = [3, 2, 4];
const target2 = 6;
console.log(`Example 2: nums = [${example2}], target = ${target2}`)
console.log("Output:", twoSum(example2, target2))

// Example 3
const example3 = [3, 3];
const target3 = 6;
console.log(`Example 3: nums = [${example3}], target = ${target3}`)
console.log("Output:", twoSum(example3, target3))

File: 01_Array_And_Hashing/two_sum.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # create hashmap to add number & index 
        prevMap = {}  # val -> index

        # iterate through list using enumerate -> returns index,value
        for i, n in enumerate(nums):
            # calculate difference (tarsget minus index value)
            diff = target - n
            # check if difference is already in hashmap
            if diff in prevMap:
                return [prevMap[diff], i]     # return pair of indecies
            # update hashmap
            prevMap[n] = i


# implementation
answer = Solution()
print("Leetcode 1: Two Sum\n")

# Example 1
example_1 = [2, 7, 11, 15]
target_1 = 9
print(f"Example 1: nums = {example_1}, target = {target_1}")
print("Output:", answer.twoSum(example_1, target_1))

# Example 2
example_2 = [3, 2, 4]
target_2 = 6
print(f"Example 1: nums = {example_2}, target = {target_2}")
print("Output:", answer.twoSum(example_2, target_2))

# Example 3
example_3 = [3, 3]
target_3 = 6
print(f"Example 1: nums = {example_3}, target = {target_1}")
print("Output:", answer.twoSum(example_3, target_3))